src/Toolchain.re

```diff
@@ -0,0 +1,759 @@
+                 switch (Json.parse(manifest)) {
+                 | Some(json) =>
+                   if (Utils.propertyExists(json, "dependencies")
+                       || Utils.propertyExists(json, "devDependencies")) {
+                     Some((Esy, projectRoot)) |> Js.Promise.resolve;
+                   } else {
+                     None |> Js.Promise.resolve;
+                   }
+
+                 | None =>
+                   Js.log3(
+                     "Invalid JSON file found. Ignoring...",
+                     manifest,
+                     manifestFile,
+                   );
+                   None |> Js.Promise.resolve;
+                 }
+               })
+          | ".opam" =>
+            Fs.readFile(manifestFile)
+            |> Js.Promise.then_(
+                 fun
+                 | "" => None |> Js.Promise.resolve
+                 | _ => Some((Opam, projectRoot)) |> Js.Promise.resolve,
+               )
+
+          | _ => None |> Js.Promise.resolve
+          };
+        };
+
+    let lookup = projectRoot => {
+      Fs.readDir(Fpath.toString(projectRoot))
+      |> Js.Promise.then_(
+           fun
+           | Error(msg) => Js.Promise.resolve(Error(msg))
+           | Ok(files) => {
+               files
+               |> Js.Array.map(parse(projectRoot))
+               |> Js.Promise.all
+               |> Js.Promise.then_(l =>
+                    Ok(
+                      Js.Array.reduce(
+                        (acc, x) =>
+                          Js.Array.concat(
+                            acc,
+                            switch (x) {
+                            | Some(x) => Array.fromList([x])
+                            | None => Array.empty
+                            },
+                          ),
+                        Array.empty,
+                        l,
+                      )
+                      |> Array.toList
+                      |> PackageManagerSpecTupleSet.of_list,
+                    )
+                    |> Js.Promise.resolve
+                  );
+             },
+         );
+    };
+  };
+};
+
+/* We declare two types with identical structure to differentiate state: t is the toolchain ready for consumption. resources is the toolchain that could need setup (and may fail) */
+
+type t = {
+  spec: PackageManager.spec,
+  projectRoot: Fpath.t,
+};
+
+type resources = {
+  spec: PackageManager.spec,
+  projectRoot: Fpath.t,
+};
+
+let init = (~env, ~folder) => {
+  let projectRoot = Fpath.ofString(folder);
+
+  Js.Promise.all2((
+    PackageManager.available(~env),
+    PackageManager.alreadyUsed(projectRoot)
+    |> Js.Promise.then_(
+         fun
+         | Ok([]) =>
+           PackageManager.Manifest.lookup(projectRoot)
+           |> okThen(x =>
+                x
+                |> PackageManager.PackageManagerSpecTupleSet.elements
+                |> (
+                  fun
+                  | [] => Error("TODO: global toolchain")
```

Comment:
What is this todo for?

Reply:
This TODO is for cases when an open file/buffer does not have any associated .merlin files since it is not a part of any project. We can reload on globally installed toolchain, but this hasn't been tested or tried